### PR TITLE
RGB <-> YCbCr

### PR DESCRIPTION
**Pull request by justanhduc**
This is my implementation of RGB <-> YCbCr. I am wondering why currently there isn't any test on GPU?

### CODE REVIEW

File: kornia/color/ycbcr.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    """
+
+    def __init__(self) -> None:
+        super(YcbcrToRgb, self).__init__()
+
+    def forward(self, image: torch.Tensor) -> torch.Tensor:  # type: ignore
+        return ycbcr_to_rgb(image)
+
+
+def ycbcr_to_rgb(image):
```
Comment (edgarriba) [inline kornia/color/ycbcr.py]: typing

File: kornia/color/ycbcr.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    r"""Convert an YCbCr image to RGB
```
Comment (edgarriba) [inline kornia/color/ycbcr.py]: end point

File: kornia/color/ycbcr.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    if len(image.shape) < 3 or image.shape[-3] != 3:
+        raise ValueError("Input size must have a shape of (*, 3, H, W). Got {}"
+                         .format(image.shape))
+
+    y = image[..., 0, :, :]
+    cb = image[..., 1, :, :]
+    cr = image[..., 2, :, :]
+
+    delta = .5
```
Comment (edgarriba) [inline kornia/color/ycbcr.py]: typing

File: kornia/color/ycbcr.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    r: torch.Tensor = y + 1.403 * (cr - delta)
+    g: torch.Tensor = y - .714 * (cr - delta) - .344 * (cb - delta)
```
Comment (edgarriba) [inline kornia/color/ycbcr.py]: would create tmp vars to not duplicate the `(cr - delta)` and `(cb - delta)`

File: kornia/color/ycbcr.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    b: torch.Tensor = y + 1.773 * (cb - delta)
+    return torch.stack((r, g, b), -3)
+
+
+class RgbToYcbcr(nn.Module):
+    r"""Convert image from RGB to YCnCr
```
Comment (edgarriba) [inline kornia/color/ycbcr.py]: `Convert image from RGB to YCbCr.`

File: kornia/color/ycbcr.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    """
+
+    def __init__(self) -> None:
+        super(RgbToYcbcr, self).__init__()
+
+    def forward(self, image: torch.Tensor) -> torch.Tensor:  # type: ignore
+        return rgb_to_ycbcr(image)
+
+
+def rgb_to_ycbcr(image):
```
Comment (edgarriba) [inline kornia/color/ycbcr.py]: typing

File: kornia/color/ycbcr.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    if len(image.shape) < 3 or image.shape[-3] != 3:
+        raise ValueError("Input size must have a shape of (*, 3, H, W). Got {}"
+                         .format(image.shape))
+
+    r: torch.Tensor = image[..., 0, :, :]
+    g: torch.Tensor = image[..., 1, :, :]
+    b: torch.Tensor = image[..., 2, :, :]
+
+    delta = .5
```
Comment (edgarriba) [inline kornia/color/ycbcr.py]: typing

File: test/color/test_ycbcr.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+class TestRgbToYcbcr:
+
+    def test_rgb_to_ycbcr(self):
+
+        data = torch.rand(3, 5, 5)
+
+        # OpenCV
+        data_cv = data.numpy().transpose(1, 2, 0).copy()
+        expected = cv2.cvtColor(data_cv, cv2.COLOR_RGB2YCrCb)
```
Comment (edgarriba) [inline test/color/test_ycbcr.py]: not sure here what makes sense to do, see @priba latest PR. He hard-coded expected values

**Comment by edgarriba** [comment]
@justanhduc please rebase from master, a recent massive PR was merged including tests for gpu. You can run it with `make test-gpu`

**Comment by justanhduc** [comment]
@edgarriba just to make sure, i should pull changes first and rebase onto master right?

**Comment by edgarriba** [comment]
@justanhduc yes, `git pull --rebase origin master`. Be aware that you might enter into conflicts
https://www.git-tower.com/learn/git/faq/solve-merge-conflicts

File: kornia/color/ycbcr.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        - image: :math:`(*, 3, H, W)`
+        - output: :math:`(*, 3, H, W)`
+
+    Examples::
+
+        >>> import torch
+        >>> import kornia
+        >>> input = torch.rand(2, 3, 4, 5)
+        >>> ycbcr = kornia.color.RgbToYcbcr()
+        >>> output = ycbcr(input)  # 2x3x4x5
```
Comment (Borda) [inline kornia/color/ycbcr.py]: why not "print" the shape and use it as a test?

File: kornia/color/ycbcr.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    if not torch.is_tensor(image):
+        raise TypeError("Input type is not a torch.Tensor. Got {}".format(
+            type(image)))
+
+    if len(image.shape) < 3 or image.shape[-3] != 3:
+        raise ValueError("Input size must have a shape of (*, 3, H, W). Got {}"
+                         .format(image.shape))
+
+    r: torch.Tensor = image[..., 0, :, :]
```
Comment (Borda) [inline kornia/color/ycbcr.py]: I guess this is not supported in older python 3.x versions...

File: test/color/test_ycbcr.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        y = result[0, :, :]
+        cb = result[1, :, :]
+        cr = result[2, :, :]
+
+        assert_allclose(y, y_expected)
+        assert_allclose(cb, cb_expected)
+        assert_allclose(cr, cr_expected)
+
+    def test_batch_rgb_to_ycbcr(self):
+        data = torch.tensor([[[0.7925, 0.8704, 0.2771, 0.3279, 0.0193],
```
Comment (Borda) [inline test/color/test_ycbcr.py]: why not randomly generated with fixed/frozen seed?

File: kornia/color/ycbcr.py (hunk 10 above)
Comment (ducha-aiki) [reply kornia/color/ycbcr.py]: kornia does not support python < 3.6 because of typing anyway

File: test/color/test_ycbcr.py (hunk 11 above)
Comment (ducha-aiki) [reply test/color/test_ycbcr.py]: What are the benefits of such approach?

Comment (Borda) [reply test/color/test_ycbcr.py]: that you do not need to insert it manually and it saves space...

**Comment by justanhduc** [comment]
@edgarriba i pulled and rebased. should i just push to the current branch?

File: test/color/test_ycbcr.py (hunk 11 above)
Comment (edgarriba) [reply test/color/test_ycbcr.py]: @Borda it also comes from the motivation of testing against results obtained with known frameworks such as opencv

Comment (Borda) [reply test/color/test_ycbcr.py]: well, what does it mean obtained results?

Comment (edgarriba) [reply test/color/test_ycbcr.py]: as far as I understood this hardcoded values come from using the matched function in opencv

**Comment by edgarriba** [comment]
@justanhduc resolved conflict throw web

File: test/color/test_ycbcr.py (hunk 11 above)
Comment (Borda) [reply test/color/test_ycbcr.py]: sounds like magic and guessing, then I think that it should be written as a comment in the code

File: kornia/color/ycbcr.py (hunk 9 above)
Comment (edgarriba) [reply kornia/color/ycbcr.py]: @justanhduc doctest plugin is not enabled. Would be great though to have such functionality implemented. You know, PRs are more than welcomed  or please open ticket asking for that functionality :)

File: test/color/test_ycbcr.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+import numpy as np
+
+import torch
+from torch.autograd import gradcheck
+from torch.testing import assert_allclose
+
+
+class TestRgbToYcbcr:
+
+    def test_rgb_to_ycbcr(self):
```
Comment (edgarriba) [inline test/color/test_ycbcr.py]: @justanhduc check da793cd48e17eca7e78b042240a9464065611d9d. Cuda test is enabled now, from root run `make test-gpu` or alternatively `pytest  test/color/test_ycbcr.py --typetest cuda`


```python
from test.common import device

def test_rgb_to_ycbcr(self, device):
    data = torch.tensot(...).to(device)

```
